packages/react-router/lib/hooks.tsx

```diff
@@ -694,18 +697,66 @@ export function _renderMatches(
     );
   }
 
+  // If we're in a partial hydration mode, detect if we need to render down to
+  // a given Fallback while we load the rest of the hydration data
+  let renderFallback = false;
+  let fallbackIndex = -1;
+  if (future && future.v7_partialHydration) {
+    for (let i = 0; i < renderedMatches.length; i++) {
+      let match = renderedMatches[i];
+      // Track the deepest fallback up until the first route without data
+      if (match.route.Fallback || match.route.fallbackElement) {
+        fallbackIndex = i;
+      }
+      if (
+        match.route.loader &&
+        match.route.id &&
+        dataRouterState?.loaderData[match.route.id] === undefined
+      ) {
+        renderFallback = true;
+        // We found the first route without data, slice to here and break the loop
+        renderedMatches = renderedMatches.slice(
+          0,
+          Math.min(renderedMatches.length, i + 1)
+        );
+        break;
+      }
+    }
+  }
+
   return renderedMatches.reduceRight((outlet, match, index) => {
-    let error = match.route.id ? errors?.[match.route.id] : null;
-    // Only data routers handle errors
+    // Only data routers handle errors/fallbacks
+    let error: any;
+    let fallback = false;
     let errorElement: React.ReactNode | null = null;
+    let fallbackElement: React.ReactNode | null = null;
     if (dataRouterState) {
+      error = errors && match.route.id ? errors[match.route.id] : null;
       errorElement = match.route.errorElement || defaultErrorElement;
+
+      if (renderFallback) {
```

Comment:
Determine if we should be rendering a fallback element at this location in the tree